### PR TITLE
new play : countries-statics

### LINKED ISSUE #392: [Add a Play]: countries-statics
Thank You! 

## What's the Play about?
This play shows the country's information around the world. Information such as name, area, demographics, capital currencies, etc. You can search for a specific country or short by region. 

## What ReactJS concept will be used to create this play?
You learn to work with useState, use effect hook and rest API data fetching techniques. 

## Are you willing to take it up for implementation?
Yes

### PR DESCRIPTION
**Pull request by Deepak8717**
> First thing, PLEASE READ THIS: [ReactPlay Code Review Checklist](https://github.com/reactplay/react-play/wiki/ReactPlay-Code-Review-Checklist)

# Description
A new play that shows countries' demographics around the world. 

Fixes #392 

## Type of change

Please delete options that are not relevant.
- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. 

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by atapas** [comment]
@Deepak8717 is it ready for review or work in progress?

**Comment by Deepak8717** [comment]
@atapas it's work in progress. I really want to have an interactive world map while countries statics and the D3 map has a learning curve which is why it's taking time.

**Comment by koustov** [comment]
> @atapas it's work in progress. I really want to have an interactive world map while countries statics and the D3 map has a learning curve which is why it's taking time.

@Deepak8717 can you move the PR to draft until it's not ready for review?

File: src/meta/play-meta.js [hunk 1]
```diff
@@ -27,7 +27,8 @@ import {
   NavBar,
```
Comment (atapas) [inline src/meta/play-meta.js:27]: This play was created with the old flow. This file shouldn't be committed anymore

File: package.json [hunk 2]
```diff
@@ -33,6 +33,7 @@
     "react-scripts": "5.0.0",
     "react-shimmer-effect": "^1.0.9",
     "react-twitter-widgets": "^1.11.0",
+    "react-simple-maps": "^2.3.0",
```
Comment (atapas) [inline package.json:36]: Can you mention this library in your play's readme like the name, link, and what it does?

File: src/plays/index.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// Do not modify the content of this file
```
Comment (atapas) [inline src/plays/index.js:1]: Do not commit this file also. In the new flow, this file shouldn't be modified.

File: src/plays/countries-statics/countriesStatics.css [hunk 4]
```diff
@@ -0,0 +1 @@
+/* enter stlyes here */
```
Comment (atapas) [inline src/plays/countries-statics/countriesStatics.css:1]: Do we need this? I see it as empty.

File: src/plays/countries-statics/Readme.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# countries-statics
+
+Fetch country specific statistic such as demographics by single click on map.
```
Comment (atapas) [inline src/plays/countries-statics/Readme.md:3]: This is a great play and an advanced one. Please provide some more technical information about it here.

File: src/plays/countries-statics/Country.jsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  const [isloading, setIsLoading] = useState(false);
+  const getCountry = () => {
+    setIsLoading(true);
+    fetch(`https://restcountries.com/v3.1/alpha/${activeGeo}`)
+      .then((response) => response.json())
+      .then((data) => {
+        setIsLoading(false);
+        setCountry(data);
+      })
+      .catch((error) => console.log(error));
```
Comment (atapas) [inline src/plays/countries-statics/Country.jsx:14]: You should take care of the loading in error too.

Also, show the error message on UI, not just on the console.

File: src/plays/countries-statics/CountriesGrid.jsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  const [isLoading, setIsLoading] = useState(true);
+  const getCountries = () => {
+    setIsLoading(true);
+    fetch(`https://restcountries.com/v3.1/all`)
+      .then((response) => response.json())
+      .then((data) => {
+        setIsLoading(false);
+        setCounteries(data);
+      })
+      .catch((error) => console.log(error));
```
Comment (atapas) [inline src/plays/countries-statics/CountriesGrid.jsx:15]: Handling error should handle loading too.

**Review by atapas — CHANGES_REQUESTED** [review]
@Deepak8717 Review comments

**Comment by atapas** [comment]
@Deepak8717 

There is no cover for the play

image

**Comment by atapas** [comment]
@Deepak8717 You can make this style better

image

**Comment by atapas** [comment]
@Deepak8717 please access the preview URL on mobile and see if the play works according to the expectations.

Here is the preview URL: https://react-play-git-fork-deepak8717-new-countries-statics-atapas.vercel.app/

I do not find the description when tapping on the map. It happens on iPhones. PLease confirm from your side too.

File: src/plays/countries-statics/Readme.md (hunk 5 above)
Comment (Deepak8717) [reply src/plays/countries-statics/Readme.md:3]: I don't want the description to be long so I provided additional details in ## Implementation Details

**Comment by Deepak8717** [comment]
@atapas all comments were taken care of except the index.js and cover image. I think we can have a cover image once the play is completed.

**Comment by atapas** [comment]
> > @atapas all comments were taken care of except the index.js and cover image. I think we can have a cover image once the play is completed.
> 
> We won’t make the play public(merging this PR) without the cover image. It won’t look good. Let’s create one.

Yes `index.js` shouldn't be committed too.

**Comment by Deepak8717** [comment]
@atapas please again

**Comment by koustov** [comment]
Marvelous... I loved this play. Great work @Deepak8717

**Comment by atapas** [comment]
Resolve merge conflicts @Deepak8717

**Review by atapas — DISMISSED** [review]
LGTM

File: src/plays/countries-statics/SearchAndFilter.jsx [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import React from "react";
+import { BsSearch } from "react-icons/bs";
+
+export default function SearchAndFilter() {
```
Comment (koustov) [inline src/plays/countries-statics/SearchAndFilter.jsx:4]: I don't think this is wired up. can you remove and add back when you are going to introduce search?

**Review by koustov — CHANGES_REQUESTED** [review]
few small comments.

**Comment by Deepak8717** [comment]
@atapas @koustov Please check

**Comment by atapas** [comment]
> @atapas @koustov Please check

@Deepak8717 have you taken care of @koustov's comments? Then let him finish it.

**Comment by atapas** [comment]
@Deepak8717 

Build fails

Failed to compile.
--
12:10:15.135 |  
12:10:15.135 | Module not found: Error: Can't resolve './SearchAndFilter' in '/vercel/path0/src/plays/countries-statics'
12:10:15.135 |  
12:10:15.135 |  
12:10:15.223 | Error: Command "CI= npm run build && npm run snap" exited with 1

File: src/plays/countries-statics/Country.jsx [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        return response.json();
+      })
+      .then((data) => {
+        setIsLoading(false);
+        setErrorInfo(null);
+        setCountry(data[0]);
+      })
+      .catch((e) => {
+        setIsLoading(false);
+        setCountry([]);
```
Comment (koustov) [inline src/plays/countries-statics/Country.jsx:23]: This should be `setCountry()` as in downstream process you are referring `country && (`. An empty array will pass the condition during rendering

**Review by koustov — CHANGES_REQUESTED** [review]
single comment

**Review by koustov — APPROVED** [review]
lgtm

**Pull request merged**